File: lib/data/model/quiz.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'quiz.freezed.dart';
part 'quiz.g.dart';

@freezed
class Quiz with _$Quiz {
  const Quiz._();

  const factory Quiz({
    required String id,
    required String articleId,
    required String title,
    required String statement,
    required bool isFavorite,
    required List<String> options,
    required List<int> answers,
    required String explanation,
  }) = _Quiz;

  factory Quiz.fromJson(Map<String, dynamic> json) => _$QuizFromJson(json);

  static Quiz mock() {
    return const Quiz(
      id: '1',
      articleId: '1',
      title: '透明UIを作る',
      statement: '透明UIとは何ですか',
      isFavorite: false,
      options: [
        'UI要素が見えないようにするテクニック',
        '背景が透明なUIデザイン',
        'ユーザーが透明なUI要素を操作できるテクニック',
      ],
      answers: [0],
      explanation: '透明UIは、UI要素が見えないようにするテクニックを指します。',
    );
  }
}

File: functions/src/types/quiz.ts
import * as admin from "firebase-admin";

export type Quiz = {
  id: string;
  articleId: string;
  title: string;
  statement: string;
  isFavorite: boolean;
  options: string[];
  answers: string[];
  explanation: string;
};

export const quizConverter = {
  toFirestore(quiz: Quiz): admin.firestore.DocumentData {
    return {
      id: quiz.id,
      articleId: quiz.articleId,
      title: quiz.title,
      statement: quiz.statement,
      isFavorite: quiz.isFavorite,
      options: quiz.options,
      answers: quiz.answers,
      explanation: quiz.explanation,
    };
  },
  fromFirestore(snapshot: admin.firestore.QueryDocumentSnapshot): Quiz {
    const data = snapshot.data();

    return {
      id: data.id,
      articleId: data.articleId,
      title: data.title,
      statement: data.statement,
      isFavorite: data.isFavorite,
      options: data.options,
      answers: data.answers,
      explanation: data.explanation,
    };
  },
};
